awx/main/access.py

```diff
@@ -552,37 +552,39 @@ def can_unattach(self, obj, sub_obj, relationship, *args, **kwargs):
 
 class OAuth2ApplicationAccess(BaseAccess):
     '''
-    I can read, change or delete OAuth applications when:
+    I can read, change or delete OAuth 2 applications when:
      - I am a superuser.
      - I am the admin of the organization of the user of the application.
-     - I am the user of the application.
-    I can create OAuth applications when:
+     - I am a user in the organization of the application.
+    I can create OAuth 2 applications when:
      - I am a superuser.
-     - I am the admin of the organization of the user of the application.
+     - I am the admin of the organization of the organization of the application.
     '''
 
     model = OAuth2Application
     select_related = ('user',)
 
     def filtered_queryset(self):
-        accessible_users = User.objects.filter(
-            pk__in=self.user.admin_of_organizations.values('member_role__members')
-        ) | User.objects.filter(pk=self.user.pk)
-        return self.model.objects.filter(user__in=accessible_users)
+        return self.model.objects.filter(organization__in=self.user.organizations)
 
     def can_change(self, obj, data):
-        return self.can_read(obj)
-
+        if obj.organization in self.user.admin_of_organizations or self.user.is_superuser:
+            return True
+        else: 
+            return False
```

Comment:
This is the case I was talking about where you need special logic if they are _changing_ the organization field, since it looks like that's still allowed by views/serializers.

You can copy the pattern in other places for `self.check_related`, and it shouldn't be too bad.

Reply:
This is fixed now and raises a helpful error message:
```
{
    "detail": "Cannot change organization without admin access to both organizations."
}```